File: src/main/java/ting/leetcode/java/hashtable/RansomNote.java
package ting.leetcode.java.hashtable;

public class RansomNote {
    public boolean canConstruct(String ransomNote, String magazine) {
        int[] cache = new int[26];
        for (char c : ransomNote.toCharArray()) {
            cache[c - 'a']++;
        }

        for (char c : magazine.toCharArray()) {
            cache[c - 'a']--;
        }

        for (int i : cache) {
            if (i > 0) return false;
        }

        return true;
    }
}

File: kotlin-solution/src/leetcode/beginnerguide/RansomNote.kt
package leetcode.beginnerguide

class RansomNote {
    fun canConstruct(ransomNote: String, magazine: String): Boolean {
        val ransomNoteList = ransomNote.toMutableList()
        magazine.forEach {
            ransomNoteList.remove(it)
            if (ransomNoteList.isEmpty())
                return true
        }
        return false
    }

    fun canConstructWithMap(ransomNote: String, magazine: String): Boolean {
        val letterCount = hashMapOf<Char, Int>()
        magazine.forEach {
            if (letterCount[it] == null) {
                letterCount[it] = 0
            } else {
                letterCount[it] = letterCount[it]!! + 1
            }
        }
        ransomNote.toCharArray().forEach {
            if (letterCount[it] == null || letterCount[it] == 0) {
                return false
            }
            if (letterCount[it] != null) {
                letterCount[it] = letterCount[it]!! - 1
            }
        }

        return true
    }
}
